internal/factsengine/gatherers/saplocalhostresolver.go

```diff
@@ -0,0 +1,177 @@
+package gatherers
+
+import (
+	"encoding/json"
+	"net"
+	"path/filepath"
+	"regexp"
+
+	probing "github.com/prometheus-community/pro-bing"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/afero"
+	"github.com/trento-project/agent/internal/core/sapsystem"
+	"github.com/trento-project/agent/pkg/factsengine/entities"
+)
+
+const (
+	SapLocalhostResolverGathererName = "saplocalhost_resolver"
+)
+
+// nolint:gochecknoglobals
+var (
+	hostnameRegexCompiled = regexp.MustCompile(`(.+)_(.+)_(.+)`) // <SID>_<InstanceNumber>_<Hostname>
+	regexSubgroupsCount   = 4
+)
+
+// nolint:gochecknoglobals
+var (
+	SapLocalhostResolverHostnameResolutionError = entities.FactGatheringError{
+		Type:    "saplocalhost_resolver-resolution-error",
+		Message: "error resolving hostname",
+	}
+	SapLocalhostResolverGathererDecodingError = entities.FactGatheringError{
+		Type:    "saplocalhost_resolver-decoding-error",
+		Message: "error decoding output to FactValue",
+	}
+	SapLocalhostResolverFileSystemError = entities.FactGatheringError{
+		Type:    "saplocalhost_resolver-file-system-error",
+		Message: "error reading the sap profiles file system",
+	}
+)
+
+//go:generate mockery --name=HostnameResolver
+type HostnameResolver interface {
+	LookupHost(host string) ([]string, error)
+}
+
+//go:generate mockery --name=HostPinger
+type HostPinger interface {
+	Ping(host string) bool
+}
+
+type SapLocalhostResolverGatherer struct {
+	fs afero.Fs
+	hr HostnameResolver
+	hp HostPinger
+}
+
+type ResolvabilityDetails struct {
+	Hostname     string   `json:"hostname"`
+	Addresses    []string `json:"addresses"`
+	InstanceName string   `json:"instance_name"`
+	Reachability bool     `json:"reachability"`
+}
+
+type Resolver struct{}
+
+type Pinger struct{}
+
+func (r *Resolver) LookupHost(host string) ([]string, error) {
+	return net.LookupHost(host)
+}
+
+func (p *Pinger) Ping(host string) bool {
+	pinger, err := probing.NewPinger(host)
+	if err != nil {
+		return false
+	}
+	pinger.Count = 3
+	err = pinger.Run()
+	return err == nil
+}
+
+func NewDefaultSapLocalhostResolverGatherer() *SapLocalhostResolverGatherer {
+	return NewSapLocalhostResolver(afero.NewOsFs(), &Resolver{}, &Pinger{})
+}
+
+func NewSapLocalhostResolver(fs afero.Fs, hr HostnameResolver, hp HostPinger) *SapLocalhostResolverGatherer {
+	return &SapLocalhostResolverGatherer{fs: fs, hr: hr, hp: hp}
+}
+
+func (r *SapLocalhostResolverGatherer) Gather(factsRequests []entities.FactRequest) ([]entities.Fact, error) {
+	facts := make([]entities.Fact, 0)
+
+	details, err := r.getInstanceHostnameDetails()
+	if err != nil {
+		log.Error(err)
+		return nil, err
+	}
+
+	for _, factReq := range factsRequests {
+		var fact entities.Fact
+		factValue, err := mapReachabilityDetailsToFactValue(details)
```

Comment:
We don't need to run `mapReachabilityDetailsToFactValue` within the loop. It uses always the same argument.
We should run it outside